internal/luks2/metadata.go

```diff
@@ -0,0 +1,874 @@
+// only be libcryptsetup) will write well-formed JSON metadata. Corruption of the JSON metadata outside
+// of modifications by libcryptsetup will be detected by the checksum verification.
+//
+// Note that this function does not attempt recovery of either header in the event that one of the
+// headers is not valid - we leave this to libcryptsetup, which happens automatically on any
+// cryptsetup or systemd-cryptsetup invocation. This package does not directly perform any
+// modifications to the header.
+//
+// This function requires an advisory shared lock on the LUKS container associated with the
+// specified path. If the mode parameter is LockModeBlocking, this function will block until the
+// lock can be obtained. If the mode parameter is LockModeNonBlocking, a wrapped syscall.Errno
+// error with the value of syscall.EWOULDBLOCK will be returned if the lock can not be obtained.
+func DecodeHdr(path string, lockMode LockMode) (*HdrInfo, error) {
+	releaseLock, err := acquireSharedLock(path, lockMode)
+	if err != nil {
+		return nil, xerrors.Errorf("cannot acquire shared lock: %w", err)
+	}
+	defer releaseLock()
+
+	f, err := os.Open(path)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	// Try to decode and check the primary header
+	primaryHdr, primaryJSON, primaryErr := decodeAndCheckHeader(f, 0, true)
+
+	var secondaryHdr *binaryHdr
+	var secondaryJSON []byte
+	if primaryErr != nil {
+		// No valid primary header. Try to decode and check a secondary header from one of the
+		// well known offsets.
+		for _, off := range []int64{0x4000, 0x8000, 0x10000, 0x20000, 0x40000, 0x80000, 0x100000, 0x200000, 0x400000} {
+			secondaryHdr, secondaryJSON, err = decodeAndCheckHeader(f, off, false)
+			if err == nil {
+				break
+			}
+		}
+	} else {
+		// Try to decode and check the secondary header immediately after the primary header.
+		secondaryHdr, secondaryJSON, _ = decodeAndCheckHeader(f, int64(primaryHdr.HdrSize), false)
+	}
+
+	var hdr *binaryHdr
+	var jsonData []byte
+	switch {
+	case primaryHdr != nil && secondaryHdr != nil:
+		// Both headers are valid
+		hdr = primaryHdr
+		jsonData = primaryJSON
+		if secondaryHdr.SeqId > primaryHdr.SeqId {
+			// The secondary header is newer than the primary header, so use that
+			hdr = secondaryHdr
+			jsonData = secondaryJSON
+		}
+	case primaryHdr != nil:
+		// We only have a valid primary header
+		hdr = primaryHdr
+		jsonData = primaryJSON
+	case secondaryHdr != nil:
+		// We only have a valid secondary header
+		hdr = secondaryHdr
+		jsonData = secondaryJSON
+	default:
+		// No valid headers :(
+		return nil, xerrors.Errorf("no valid header found, error from decoding primary header: %w", primaryErr)
+	}
+
+	info := &HdrInfo{
+		HdrSize: hdr.HdrSize,
+		Label:   hdr.Label.String()}
+
+	if err := json.NewDecoder(bytes.NewReader(jsonData)).Decode(&info.Metadata); err != nil {
+		return nil, err
```

Comment:
from the description an error here seems fairly unexpected? it should be prefixed with a bit more contextual info I think than just the bare encoding/json error

Reply:
I've added context to the error now. I've also moved the JSON decoding to before the header selection to make the code a bit more robust (if JSON decoding fails for one header, it provides an opportunity to fallback to the other header if that's ok).